crates/containerd-shim-wasm/src/container/engine.rs

```diff
@@ -53,12 +54,14 @@ pub trait Engine: Clone + Send + Sync + 'static {
         &["application/vnd.bytecodealliance.wasm.component.layer.v0+wasm"]
     }
 
-    /// Precompiles a module that is in the WASM OCI layer format
-    /// This is used to precompile a module before it is run and will be called if can_precompile returns true.
+    /// Precompile passes supported OCI layers to engine for compilation
+    /// This is used to precompile the layers before they are run and will be called if `can_precompile` returns `true`.
     /// It is called only the first time a module is run and the resulting bytes will be cached in the containerd content store.  
-    /// The cached, precompiled module will be reloaded on subsequent runs.
-    fn precompile(&self, _layers: &[Vec<u8>]) -> Result<Vec<u8>> {
-        bail!("precompilation not supported for this runtime")
+    /// The cached, precompiled layers will be reloaded on subsequent runs.
+    /// The runtime is expected to return the same number of layers passed in, if the layer cannot be precompiled it should return `None` for that layer.
+    /// In some edge cases it is possible that the layers may already be precompiled and None should be returned in this case.
+    fn precompile(&self, _layers: &[WasmLayer]) -> Result<Vec<Option<Vec<u8>>>> {
```

Comment:
nit: type alias for `Vec<Option<Vec<u8>>>>`, perhaps `PrecompiledWasmLayers`?

Reply:
One nit to counter the nit with: not everything returned is a precompiled layer, which is why there is an option wrapping, so it feels like a confusing name.

Reply:
Or maybe `Vec<Option<WasmLayer>>`?

Reply:
`WasmLayer` is already a structure. The API was that at one point but was changed to just return the layer content rather than the content and config because the original config is used as the source of truth. I think we should maybe leave it as is or `PrecompiledLayer` could work, though i prefer not using types to alias something as small as `Vec<u8>`